### PR TITLE
adding cel for allow-privilege-escalation

### PR DESCRIPTION
**Pull request by maxsmythe**
**What this PR does / why we need it**:

This converts K8sAllowPrivilegeEscalation to use CEL in addition to Rego.

**Which issue(s) does this PR fix** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #541 

**Special notes for your reviewer**:

### CODE REVIEW

File: library/pod-security-policy/allow-privilege-escalation/template.yaml [hunk 1]
```diff
@@ -36,6 +36,35 @@ spec:
...
+          - name: initContainers
+            expression: 'has(object.spec.initContainers) ? object.spec.initContainers : []'
+          - name: ephemeralContainers
+            expression: 'has(object.spec.ephemeralContainers) ? object.spec.ephemeralContainers : []'
+          - name: exemptImages
+            expression: |
+              !has(variables.params.exemptImages) ? [] :
+                (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                  // we can use the regex substitution because docker images do not allow "." as a valid character
+                  variables.params.exemptImages.exists(exemption, container.image == exemption || (exemption.endsWith("*") && string(container.image).matches("^" + string(variables.params.exemptImages).replace("*", ".*") + "$")))
```
Comment (ritazh) [inline library/pod-security-policy/allow-privilege-escalation/template.yaml:54]: I tested with this and got this error `composited variable "badContainers" fails to evaluate: composited variable "exemptImages" fails to evaluate: no such overload: string(list)`
```
 parameters:
    exemptImages: ["*"]
```

File: library/pod-security-policy/allow-privilege-escalation/template.yaml [hunk 2]
```diff
@@ -36,6 +36,35 @@ spec:
...
+                )
+          - name: badContainers
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                !(container.image in variables.exemptImages) && (
```
Comment (ritazh) [inline library/pod-security-policy/allow-privilege-escalation/template.yaml:59]:
```suggestion
                !(container in variables.exemptImages) && (
```

made this change to make exemption work

File: library/pod-security-policy/allow-privilege-escalation/template.yaml (hunk 1 above)
Comment (maxsmythe) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml:54]: Gah, should have been `string(exemption)` -- exemptImages is a list

Comment (maxsmythe) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml]: fixed

File: library/pod-security-policy/allow-privilege-escalation/template.yaml (hunk 2 above)
Comment (maxsmythe) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml]: actually, this is wrong, we should be using the image names.

Comment (maxsmythe) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml:59]: updated to use image string

File: library/pod-security-policy/allow-privilege-escalation/template.yaml [hunk 3]
```diff
@@ -36,6 +36,42 @@ spec:
...
+                  !has(container.securityContext) ||
+                  !has(container.securityContext.allowPrivilegeEscalation) ||
+                  container.securityContext.allowPrivilegeEscalation != false
+                )
+              )
+          validations:
+          - expression: '(has(request.operation) && request.operation == "UPDATE") || size(variables.badContainers) == 0'
```
Comment (JaydipGabani) [inline library/pod-security-policy/allow-privilege-escalation/template.yaml:73]: wouldn't this return `true` for all `Update` requrests regardless if the updates have `badContainers` or not and get denied? we rather want something like - `(has(request.operation) && request.operation != "UPDATE") && size(variables.badContainers) != 0` - if the requrest in not `UPDATE` and there are `badContainers`, throw violation.

**Review by JaydipGabani — APPROVED** [review]

File: library/pod-security-policy/allow-privilege-escalation/template.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (JaydipGabani) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml:73]: nvm, I think I got it other way around. You want `expression` to be `false` to throw violation. In that case the expression is correct.

File: library/pod-security-policy/allow-privilege-escalation/template.yaml [hunk 4]
```diff
@@ -36,6 +36,42 @@ spec:
...
+              !has(variables.params.exemptImages) ? [] :
+                variables.params.exemptImages.filter(image, image.endsWith("*")).map(image, string(image).replace("*", ""))
+          - name: exemptImageExplicit
+            expression: |
+              !has(variables.params.exemptImages) ? [] : 
+                variables.params.exemptImages.filter(image, !image.endsWith("*"))
+          - name: exemptImages
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                container in variables.exemptImageExplicit ||
```
Comment (ritazh) [inline library/pod-security-policy/allow-privilege-escalation/template.yaml:60]:
```suggestion
                container.image in variables.exemptImageExplicit ||
```
`variables.exemptImageExplicit` is a list of images, not containers.

Comment (maxsmythe) [reply library/pod-security-policy/allow-privilege-escalation/template.yaml:60]: ack, fixing and relocating code.

**Comment by maxsmythe** [comment]
bumped minor version

**Comment by ritazh** [comment]
@maxsmythe Just confirming, this PR is ready for review right? not a draft.

**Review by JaydipGabani — APPROVED** [review]
cel LGTM

**Comment by maxsmythe** [comment]
@ritazh Right, no longer a draft, but curious why checks were canceled

**Review by ritazh — APPROVED** [review]
LGTM

**Pull request merged**